### PR TITLE
TorchToTosa: torch.aten.empty.memory_format

### PR DESCRIPTION
**Pull request by flemairen6**

### CODE REVIEW

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -4303,7 +4305,7 @@ class ConvertAtenAvgPool2dOp
   }
 };
 
-// Ref: Error checking based on the Torch to LinAlg lowering
```
Comment (mgehre-amd) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:4306]: Why did you remove this?

**Review by mgehre-amd — APPROVED** [review]

**Comment by mgehre-amd** [comment]
`NewEmptyModuleLayoutIntDtype_basic` crashes for tosa

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -4614,6 +4616,96 @@ LogicalResult ConvertAtenOp<AtenSqrtOp>::matchAndRewrite(
   return success();
 }
 
+template <>
+LogicalResult ConvertAtenOp<AtenEmptyMemoryFormatOp>::matchAndRewrite(AtenEmptyMemoryFormatOp op, OpAdaptor adaptor,
```
Comment (ttjost) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:4620]: Could you fix the format here, please?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 3]
```diff
@@ -4614,6 +4616,96 @@ LogicalResult ConvertAtenOp<AtenSqrtOp>::matchAndRewrite(
...
+      return rewriter.notifyMatchFailure(
+          op, "Unsupported pin_memory, should be either None or false");
+    }
+
+    if (!op.getDevice().getType().template isa<Torch::NoneType>()) {
+      std::string device;
+      if (!matchPattern(op.getDevice(), m_TorchConstantDevice(device)))
+        return rewriter.notifyMatchFailure(
+            op, "unimplemented: device must be a constant str");
+      else if (device != "cpu")
```
Comment (ttjost) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:4639]: Do we need the `else` here?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 4]
```diff
@@ -4614,6 +4616,96 @@ LogicalResult ConvertAtenOp<AtenSqrtOp>::matchAndRewrite(
...
+        return rewriter.notifyMatchFailure(
+            op, "unimplemented: device is expected to be none or cpu");
+    }
+
+    if (!op.getLayout().getType().template isa<Torch::NoneType>()) {
+      int64_t tensorLayout;
+      if (!matchPattern(op.getLayout(), m_TorchConstantInt(&tensorLayout)))
+        return rewriter.notifyMatchFailure(
+            op, "unimplemented: layout must be a constant");
+      else if (tensorLayout != torch_upstream::Layout::Strided)
```
Comment (ttjost) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:4649]: Same here.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 5]
```diff
@@ -4614,6 +4616,96 @@ LogicalResult ConvertAtenOp<AtenSqrtOp>::matchAndRewrite(
...
+      if(maybeResultElementType->isSignedInteger(32))
+        emptyVal  = DenseIntElementsAttr::get(resultType, {0});
+      else if (maybeResultElementType->isSignlessInteger(64))
+        emptyVal  = DenseIntElementsAttr::get(resultType, {0UL});
+      else if (maybeResultElementType->isSignlessInteger(32))
+        emptyVal  = DenseIntElementsAttr::get(resultType, {0U});
+      else if (maybeResultElementType->isF64())
+        emptyVal  = DenseFPElementsAttr::get(resultType, {0.0});
+      else if (maybeResultElementType->isF32())
+        emptyVal  = DenseFPElementsAttr::get(resultType, {0.0F});
```
Comment (ttjost) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:4702]: Just for the sake of asking: did you test what happens if `maybeResultElementType` is a `bf16`?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 6]
```diff
@@ -4802,6 +4894,7 @@ class ConvertTorchToTosa : public ConvertTorchToTosaBase<ConvertTorchToTosa> {
   target.addIllegalOp<AtenOp>();                                               \
   patterns.add<ConvertAtenFillScalarOp<AtenOp>>(typeConverter, context);
     INSERT_FILL_SCALAR_PATTERN(AtenFill_ScalarOp);
+    INSERT_FILL_SCALAR_PATTERN(AtenFillScalarOp);
```
Comment (ttjost) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:4897]: Is this related to the pass?

Comment (flemairen6) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:4897]: Yes, every Empty comes with a fill, this is just legalizing the `fill` using the same implementation as the `fill_` op

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 5 above)
Comment (flemairen6) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:4702]: As I was expecting, it doesn't work (the whole reason for those `if`s is because there is a check in LLVM for the width of the element compared to the width of the type, and it doesn't cast the incoming type)

Comment (flemairen6) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:4702]: I have yet to see the case in an app, so I'm thinking about leaving it out for the moment, since it's a bit more complex to have a bfloat `0.0` for this function

Comment (mgehre-amd) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:4703]: The please add a final else branch that returns a rewriter.notifyMatchFailure when the type is not handled here.

Comment (flemairen6) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:4703]: I tried to make it work with this method, but it makes it wide use of the C API and some functions that should not be called from the outside, so for the moment I'll just mark it as a `notifyMatchFailure` and if the case appears where we need to support bfloat16 here, I'll take a deeper look into making a proper implementation for it.
